interfaces/systemd/backend.go

```diff
@@ -0,0 +1,164 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package systemd implements integration between snappy interfaces and
+// arbitrary systemd units that may be required for "oneshot" style tasks.
+package systemd
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+	sysd "github.com/snapcore/snapd/systemd"
+)
+
+// Backend is responsible for maintaining apparmor profiles for ubuntu-core-launcher.
+type Backend struct{}
+
+// Name returns the name of the backend.
+func (b *Backend) Name() string {
+	return "systemd"
+}
+
+func (b *Backend) Setup(snapInfo *snap.Info, devMode bool, repo *interfaces.Repository) error {
+	snapName := snapInfo.Name()
+	rawSnippets, err := repo.SecuritySnippetsForSnap(snapInfo.Name(), interfaces.SecuritySystemd)
+	if err != nil {
+		return fmt.Errorf("cannot obtain systemd security snippets for snap %q: %s", snapName, err)
+	}
+	snippets, err := unmarshalRawSnippetMap(rawSnippets)
+	if err != nil {
+		return fmt.Errorf("cannot unmarshal systemd snippets for snap %q: %s", snapName, err)
+	}
+	snippet, err := flattenSnippetMap(snippets)
+	if err != nil {
+		return fmt.Errorf("cannot merge systemd snippets for snap %q: %s", snapName, err)
+	}
+	content, err := renderSnippet(snippet)
+	if err != nil {
+		return fmt.Errorf("cannot render systemd snippets for snap %q: %s", snapName, err)
+	}
+	dir := dirs.SnapServicesDir
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("cannot create directory for systemd services %q: %s", dir, err)
+	}
+	glob := interfaces.SystemdServiceGlob(snapName)
+	changed, removed, errEnsure := osutil.EnsureDirState(dir, glob, content)
+	systemd := sysd.New(dirs.GlobalRootDir, &dummyReporter{})
+	// Start any new services
+	for _, service := range changed {
+		err := systemd.Start(service)
+		if err != nil {
+			logger.Noticef("cannot start service %q: %s", service, err)
+		}
+	}
+	// Stop any removed services
+	for _, service := range removed {
+		err := systemd.Stop(service, 1*time.Second)
+		if err != nil {
+			logger.Noticef("cannot stop service %q: %s", service, err)
+		}
+	}
+	// Reload systemd whenever something is removed
+	if len(removed) > 0 {
```

Comment:
This is the remove case in which we always remove all of the content. The `Setup` case was indeed fixed.